### PR TITLE
【Hackathon No.28】RFC For logcumsumexp

### PR DESCRIPTION
**Pull request by tiancaishaonvjituizi**

### CODE REVIEW

File: rfcs/APIs/20220330_api_design_for_logcumsumexp.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+在飞桨中增加 logcumsumexp API。
+
+## 3、意义
+
+飞桨将支持 logcumsumexp API。
+
+# 二、飞桨现状
+
+在飞桨中，logcumsumexp 可以通过已有的 API 组合而成：paddle.log(paddle.cumsum(paddle.exp(x)))，但这样做的数值稳定性很差。
```
Comment (luotao1) [inline rfcs/APIs/20220330_api_design_for_logcumsumexp.md:28]: > 但这样做的数值稳定性很差

是避免指数引起的溢出么？

除了数值稳定性，用python组合的话，性能方面也会较差

File: rfcs/APIs/20220330_api_design_for_logcumsumexp.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+```
+
+axis 默认为 None，表示将 tensor flatten 后再进行操作；dtype 会用于将 input tensor cast 到另一种数据类型，防止累加的溢出；exclusive 表示是否进行 exclusive prefix scan；reverse 表示是否是从尾部开始由后往前计算累加和。
+
+## 底层OP设计
+
+一定程度上可以仿照 Paddle 中已有的 cumsum op 来实现。Paddle 的 cumsum CUDA kernel 中，直接调用了 thrust 的 API `thrust::exclusive_scan` 或 `thrust::inclusive_scan` 。幸运的是 thrust 有一族 [Transformed Prefix Scans 接口](https://thrust.github.io/doc/group__transformed__prefixsums.html)，支持以任意一个满足结合律的二元运算代替 “加法” 的角色。因此 logcumsumexp 的 CUDA kernel 只要使用这族接口，将 log_add_exp 作为参与 prefix scan 的二元运算即可。
+
+但 Paddle cumsum CPU kernel 使用 eigen 作为运算库，eigen 只有普通的 cumsum 和 cumprod 接口，而不支持任意满足结合律的二元运算，因此 Paddle logcumsumexp 的 CPU Kernel 可能只能先写一个串行的 naive 版而不容易实现高效的并行。
```
Comment (luotao1) [inline rfcs/APIs/20220330_api_design_for_logcumsumexp.md:83]: 去掉`可能`，可以写这一版本写一个串行的naive版

File: rfcs/APIs/20220330_api_design_for_logcumsumexp.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+## API实现方案
+
+仍是类似于 Paddle cumsum op 即可。
+
+# 六、测试和验收的考量
+
+实现基于 NumPy 的参考实现，将 Paddle logcumsumexp kernel 的结果与参考实现对比。在 CPU 和 GPU、静态图与动态图模式，以及各种参数设置下，均与 NumPy 的结果一致。
```
Comment (luotao1) [inline rfcs/APIs/20220330_api_design_for_logcumsumexp.md:91]: 因为是新增C++ Op，请参考：
- 根据[验收标准](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/dev_guides/api_contributing_guides/api_accpetance_criteria_cn.html#id1)增加测试部分
- [cholesky_solve](https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/20220329_design_for_cholesky_solve.md)也是一个C++ Op，可以参考它的【测试和验收考量】
- 数值不稳定性，即防止溢出，需要有单测保证

既然上面提到的性能，可以增加与业内其他方案的性能对比，该部分不需要写入单测，但在提PR时可以讲自测代码贴在PR描述或评论中。

File: rfcs/APIs/20220330_api_design_for_logcumsumexp.md (hunk 1 above)
Comment (tiancaishaonvjituizi) [reply rfcs/APIs/20220330_api_design_for_logcumsumexp.md]: > 是避免指数引起的溢出么？

嗯嗯是的，在 RFC 中已补充上了说明。

> 除了数值稳定性，用python组合的话，性能方面也会较差

好的，已补充

File: rfcs/APIs/20220330_api_design_for_logcumsumexp.md (hunk 2 above)
Comment (tiancaishaonvjituizi) [reply rfcs/APIs/20220330_api_design_for_logcumsumexp.md]: 好的，已修改

File: rfcs/APIs/20220330_api_design_for_logcumsumexp.md (hunk 3 above)
Comment (tiancaishaonvjituizi) [reply rfcs/APIs/20220330_api_design_for_logcumsumexp.md]: 好的，已修改

File: rfcs/APIs/20220330_api_design_for_logcumsumexp.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+# 六、测试和验收的考量
+
+实现基于 NumPy 的参考实现，预期实现效果与 NumPy 保持一致：
+
+1. 测试API 动态图和静态图下的一致性。
+2. 测试CPU、GPU上的一致性。
+3. 测试在 fp16、fp32 和 fp64 下的一致性。
+4. 构造几个按 paddle.log(paddle.cumsum(paddle.exp(x))) 的计算方式来计算会溢出，使用变换后的 log_add_exp 计算则不会溢出的的测试用例，验证该 OP 的实现不会发生溢出。
+5. axis/exclusive/reverse/dtype 的每一个可能的取值都需要有测试用例覆盖到。
+6. 在本地对比与 TensorFlow 的性能，确保 CUDA 版本不低于 PyTorch。
```
Comment (luotao1) [inline rfcs/APIs/20220330_api_design_for_logcumsumexp.md:98]: > 在本地对比与 TensorFlow 的性能，确保 CUDA 版本不低于 PyTorch

这句话前面是TF，后面是PyTorch，是两者都要对比一下么？

Comment (tiancaishaonvjituizi) [reply rfcs/APIs/20220330_api_design_for_logcumsumexp.md]: 抱歉，是手误。已修改

**Review by luotao1 — APPROVED** [review]
LGTM

File: rfcs/APIs/20220330_api_design_for_logcumsumexp.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+仍是类似于 Paddle cumsum op 即可。
+
+# 六、测试和验收的考量
+
+实现基于 NumPy 的参考实现，预期实现效果与 NumPy 保持一致：
+
+1. 测试API 动态图和静态图下的一致性。
+2. 测试CPU、GPU上的一致性。
+3. 测试在 fp16、fp32 和 fp64 下的一致性。
+4. 构造几个按 paddle.log(paddle.cumsum(paddle.exp(x))) 的计算方式来计算会溢出，使用变换后的 log_add_exp 计算则不会溢出的的测试用例，验证该 OP 的实现不会发生溢出。
```
Comment (Yulv-git) [inline rfcs/APIs/20220330_api_design_for_logcumsumexp.md:96]: **的的**附近，可能出现了拼写错误。
